### DISCUSSION TITLE
Long duration to get the list of ble device

### DISCUSSION THREAD

**Discussion by staik9** [inthehand/32feet#403] [Q&A] [answered]
Hi,
I am using ScanForDevicesAsync method to get a list of the BLE devices that are discoverable on Windows. What I noticed is that it always takes 30sec to print me a list of the found devices. I used the filter option to only look for devices with a certain prefix and it also takes 30sec. I would appreciate if you can suggest a solution to minimize the time or give me a list in real time. 
Thank you

**Comment by peterfoot** [comment, marked as answer]
At the moment this is a limitation of the DeviceInformation.FindAllAsync which Windows implementation uses. It needs to be replaced with a  DeviceWatcher which will return results as they are found rather than all at once after a non-configurable timeout. It's on the to-do list but I'm not sure when it will be completed.

**Reply by staik9** [reply]
Thank you for your answer and clarification. Now I switched to RequestLEScanAsync method, however, the BLE device's name only gets discovered once I open the Bluetooth settings of my computer. But, when I check the MAC address of the BLE device I find it in the list I am printing in the console. What could be the problem ? Thank you in advance.

**Reply by GitBurs** [reply]
> At the moment this is a limitation of the DeviceInformation.FindAllAsync which Windows implementation uses. It needs to be replaced with a DeviceWatcher which will return results as they are found rather than all at once after a non-configurable timeout. It's on the to-do list but I'm not sure when it will be completed.

One year later and I guess this is still on the to-do list? Or even ignored, as there was only a single update in the past 12 months, while I see some people asked the same questions even in early 2023. So, not sure what's going on here, but it shouldn't be so complicated to implement the changes (tho Windows Bluetooth stack is indeed a pretty messed up thing from my experience).

EDIT: it's actually not that hard to implement BluetoothLEAdvertisementWatcher yourself. After you receive a notification from your Bluetooth device, just pass its Id to InTheHand's Bluetooth.BluetoothDevice.FromIdAsync() method. The Id should be either the one you got from BluetoothLEDevice.DeviceInformation.Id or device's actual Bluetooth address without dash or colon chars (only 12 alpha-numeric chars in total). But you can also just continue working with BluetoothLEDevice object type as it's "compatible" with InTheHand's type (it's inherited).

Note: don't use FromIdAsync() method alone (without the watcher first "confirming" the device is actually available), otherwise some weird things will happen! Like, even when your Bluetooth device is off, the method will return the BluetoothDevice object the same way as when the device is on (assuming the device was discovered once before)! Which will obviously break things down the road. The only "fix" is to disable and enable Bluetooth radio in Device Manager (and again and again when you find yourself in the same situation - which will become tiresome quite fast).
